# <div style="font-family: 'Arial'; background-color: moccasin; padding: 10px; color: Red; text-align: center;">IPL Data Analysis</div>

In [1]:
#importing the libraries to perform requried operations
import numpy as np
import pandas as pd
import plotly.express as px
import seaborn as sns

In [2]:
#loadind datasets
#1.loading the matches dataset 
matches=pd.read_csv("D:\IPL Matches 2008-2020.csv")

#2.loading the Ball-by-Ball dataset
deleviries = pd.read_csv("D:\IPL Ball-by-Ball 2008-2020.csv")

<a id="33">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; text-align: center; font-style: normal; background-color: orange; color :black;"> EDA </h1>


In [3]:
#basic imformation about the matches dataset
matches.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 816 entries, 0 to 815
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               816 non-null    int64  
 1   city             803 non-null    object 
 2   date             816 non-null    object 
 3   player_of_match  812 non-null    object 
 4   venue            816 non-null    object 
 5   neutral_venue    816 non-null    int64  
 6   team1            816 non-null    object 
 7   team2            816 non-null    object 
 8   toss_winner      816 non-null    object 
 9   toss_decision    816 non-null    object 
 10  winner           812 non-null    object 
 11  result           812 non-null    object 
 12  result_margin    799 non-null    float64
 13  eliminator       812 non-null    object 
 14  method           19 non-null     object 
 15  umpire1          816 non-null    object 
 16  umpire2          816 non-null    object 
dtypes: float64(1), i

In [4]:
#basic imformation about the ball-by-ball dataset
deleviries.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 193468 entries, 0 to 193467
Data columns (total 18 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                193468 non-null  int64 
 1   inning            193468 non-null  int64 
 2   over              193468 non-null  int64 
 3   ball              193468 non-null  int64 
 4   batsman           193468 non-null  object
 5   non_striker       193468 non-null  object
 6   bowler            193468 non-null  object
 7   batsman_runs      193468 non-null  int64 
 8   extra_runs        193468 non-null  int64 
 9   total_runs        193468 non-null  int64 
 10  non_boundary      193468 non-null  int64 
 11  is_wicket         193468 non-null  int64 
 12  dismissal_kind    9495 non-null    object
 13  player_dismissed  9495 non-null    object
 14  fielder           6784 non-null    object
 15  extras_type       10233 non-null   object
 16  batting_team      193468 non-null  obj

##### Checking Null values

In [5]:
#cheacking null values in matches dataset
matches.isna().sum()

id                   0
city                13
date                 0
player_of_match      4
venue                0
neutral_venue        0
team1                0
team2                0
toss_winner          0
toss_decision        0
winner               4
result               4
result_margin       17
eliminator           4
method             797
umpire1              0
umpire2              0
dtype: int64

In [6]:
#cheaking null values in delevaries dataset
deleviries.isna().sum()

id                       0
inning                   0
over                     0
ball                     0
batsman                  0
non_striker              0
bowler                   0
batsman_runs             0
extra_runs               0
total_runs               0
non_boundary             0
is_wicket                0
dismissal_kind      183973
player_dismissed    183973
fielder             186684
extras_type         183235
batting_team             0
bowling_team           191
dtype: int64

In [7]:
#basic information of the dataset
deleviries.describe()

id         inning           over           ball  \
count  1.934680e+05  193468.000000  193468.000000  193468.000000   
mean   7.567688e+05       1.482131       9.177027       3.615967   
std    3.060971e+05       0.499682       5.676848       1.807128   
min    3.359820e+05       1.000000       0.000000       1.000000   
25%    5.012270e+05       1.000000       4.000000       2.000000   
50%    7.292970e+05       1.000000       9.000000       4.000000   
75%    1.082628e+06       2.000000      14.000000       5.000000   
max    1.237181e+06       2.000000      19.000000       9.000000   

        batsman_runs     extra_runs     total_runs   non_boundary  \
count  193468.000000  193468.000000  193468.000000  193468.000000   
mean        1.240231       0.066414       1.306645       0.000083   
std         1.610867       0.339991       1.598802       0.009094   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         1.000000       0.000000       1.000000       0.000000   
75%         1.000000       0.000000       1.000000       0.000000   
max         6.000000       7.000000       7.000000       1.000000   

           is_wicket  
count  193468.000000  
mean        0.049078  
std         0.216031  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000

In [8]:
#basic information about the dataset
matches.describe()

id  neutral_venue  result_margin
count  8.160000e+02     816.000000     799.000000
mean   7.563496e+05       0.094363      17.321652
std    3.058943e+05       0.292512      22.068427
min    3.359820e+05       0.000000       1.000000
25%    5.012278e+05       0.000000       6.000000
50%    7.292980e+05       0.000000       8.000000
75%    1.082626e+06       0.000000      19.500000
max    1.237181e+06       1.000000     146.000000

<a id="30">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color:orange; color :black;">1) Count of matches played in each season </h1>


In [9]:
# finding out the season based on the year and creating a new column
matches['season'] = pd.DatetimeIndex(matches['date']).year
matches.head(2)

id        city        date player_of_match  \
0  335982   Bangalore  2008-04-18     BB McCullum   
1  335983  Chandigarh  2008-04-19      MEK Hussey   

                                        venue  neutral_venue  \
0                       M Chinnaswamy Stadium              0   
1  Punjab Cricket Association Stadium, Mohali              0   

                         team1                  team2  \
0  Royal Challengers Bangalore  Kolkata Knight Riders   
1              Kings XI Punjab    Chennai Super Kings   

                   toss_winner toss_decision                 winner result  \
0  Royal Challengers Bangalore         field  Kolkata Knight Riders   runs   
1          Chennai Super Kings           bat    Chennai Super Kings   runs   

   result_margin eliminator method    umpire1      umpire2  season  
0          140.0          N    NaN  Asad Rauf  RE Koertzen    2008  
1           33.0          N    NaN  MR Benson   SL Shastri    2008

In [10]:
runs_data_frame = [] #creating a empty list
print('number of matches played in each season')
# taking the unique value of season from the column
for i in matches['season'].unique(): # unique values will give the number of seasons played 
    # printing the number of matches played in each season 
    runs_data_frame.append({'season' : i , 'No_of_Matches' : len(matches[matches['season'] == i]) })
    
matches_per_season_df = pd.DataFrame(runs_data_frame)

# Display the created dataframe
matches_per_season_df.head()

number of matches played in each season


season  No_of_Matches
0    2008             58
1    2009             57
2    2010             60
3    2011             73
4    2012             74

In [11]:
#visualization of matches played in each season using ploty.express
colors = ['turquoise',] * 15
colors[5] = 'moccasin'
#loading dataset and providing the values of x and y 
fig = px.bar(data_frame=matches_per_season_df ,x=matches_per_season_df.season
                         , y=matches_per_season_df.No_of_Matches )
#giving the title to the visualization
fig.update_layout(title = 'No of Matches per season')
#providing the marker line color and and width
fig.update_traces(marker_line_color='Black' , marker_line_width=1.5,marker_color=colors)
fig.show()

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>2013</b> season has most no of matches
</p>
</div>


<a id="30">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color:orange; color :black;">2) Runs were scored in each season </h1>


In [12]:
# merging two datasets with required columns and creating a new dataset
data = matches[['season','id']].merge(deleviries , left_on= 'id' , right_on='id' , how='left').drop('id', axis=1)

In [13]:
# calculating the total runs scored in each season using groupby function
runs_data = data.groupby(['season'])['total_runs'].sum().reset_index()

#creating dataframe of total runs in each season
runs_per_season = pd.DataFrame(data=runs_data , columns=['season','total_runs'])
runs_per_season.head(2)

season  total_runs
0    2008       17937
1    2009       16320

In [14]:
# visualization of total runs in each season
colors = ['skyblue',] * 13
colors[5] = 'darkmagenta'
#providing the dataset and the values of x and y
fig = px.bar( data_frame=runs_per_season ,x = runs_per_season.season,y = runs_per_season.total_runs)
#providing the title for the visualization
fig.update_layout(title= "Runs Per Season")
#proving the marker line width , colors etc.
fig.update_traces(marker_line_color= 'Black' , marker_line_width=1.2 , 
                  marker_color=colors)


<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>2013</b> is the season in which most runs were scored
</p>
</div>


<a id="30">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color:orange; color :black;">3) Runs Scored Per match in each season </h1>


In [15]:
#calculating the avg runs per match in each season
matches_per_season_df['runs_per_match'] = runs_per_season['total_runs'] / matches_per_season_df['No_of_Matches']
matches_per_season_df

season  No_of_Matches  runs_per_match
0     2008             58      309.258621
1     2009             57      286.315789
2     2010             60      314.400000
3     2011             73      289.780822
4     2012             74      303.418919
5     2013             76      296.592105
6     2014             60      315.150000
7     2015             59      310.711864
8     2016             60      314.366667
9     2017             59      318.118644
10    2018             60      331.683333
11    2019             60      323.333333
12    2020             60      322.533333

In [16]:
#data visualization of avg runs per match in each season
#providing the dataset and the values of x and y
fig = px.line(data_frame=matches_per_season_df , x=matches_per_season_df.season ,
             y=matches_per_season_df.runs_per_match)
#providing the title to the line graph
fig.update_layout(title = 'Runs Per Match')

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>2018</b> season has highest avg runs of all time
</p>
</div>


<a id="33">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color: orange; color :black;">4) which umpire has umpired most no oF matches:</h1>


In [17]:
# calculating which umpire has umpired most no. of matches form umpire1 column
umpire_data = []  #an empty list 
#using for loop to get no of matches upired by which umpired
for umpire in matches['umpire1'].unique():
    umpire_data.append({'Umpire' : umpire , 'matches umpired' : len(matches[matches['umpire1'] == umpire])})
 
 # preapring a dataset of upire and their total matches   
umpire_data_df = pd.DataFrame(data=umpire_data)
#sorting it based on their no matches upired
umpire_data_df.sort_values(by='matches umpired' , ascending=False).head()

Umpire  matches umpired
12  HDPK Dharmasena               78
25     AK Chaudhary               56
0         Asad Rauf               51
11        M Erasmus               40
2         Aleem Dar               38

In [18]:
# calculating which umpire has umpired most no. of matches form umpire2 column
umpire_data_1 = [] #empty list
#using for loop to get no of matches upired by which umpired
for umpire in matches['umpire2'].unique():
    umpire_data_1.append({'Umpire' : umpire , 'matches umpired' : len(matches[matches['umpire2'] == umpire])})
# preapring a dataset of upire and their total matches      
umpire_data_df_1 = pd.DataFrame(data=umpire_data_1)
#sorting it based on their no matches upired
umpire_data_df_1.sort_values(by='matches umpired' , ascending=False).head()

Umpire  matches umpired
19         S Ravi               84
30  C Shamshuddin               60
15     SJA Taufel               54
34      CK Nandan               49
24      RJ Tucker               41

In [19]:
# calculating which umpire has umpired most no. of matches from umpir1 and umpire2 column 
umps = pd.concat([matches['umpire1'], matches['umpire2']]).value_counts().reset_index().rename(columns={'index': 'umpire',0: 'Matches Umpired'})
ump_top_10 = umps.head(10)

#data visulization of umpire with their no of matches umpires
colors =['deeppink']*15
colors[0] = 'darkmagenta'
fig = px.bar(data_frame=ump_top_10 ,x=ump_top_10["count"] , y=ump_top_10.umpire)
#providing the title
fig.update_layout(title= 'No of Matches Umpired by an Umpire ')
#specification of marker line
fig.update_traces(marker_line_color='black',
                  marker_line_width=1.2, opacity=1 , marker_color = colors)

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>S Ravi </b> has umpired most no of matches
</p>
</div>


<a id="33">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color: orange; color :black;">5) Most number of toss winner:</h1>


In [20]:
#calculating who has won most no of tosses
toss_winner_1 = [] #empty list
#accessing the names of teams and preapring the dataset of with their no of toss wins
for team in matches['toss_winner'].unique():
    #appending the data into the list in dictonary format
    toss_winner_1.append({ "Team": team , "Number of Toss wins" : len(matches[matches['toss_winner'] == team])})
#preparing the dataset
toss_winner_df = pd.DataFrame(data= toss_winner_1)
toss_winner_df.head(10)

Team  Number of Toss wins
0  Royal Challengers Bangalore                   87
1          Chennai Super Kings                   97
2             Rajasthan Royals                   87
3               Mumbai Indians                  106
4              Deccan Chargers                   43
5              Kings XI Punjab                   85
6        Kolkata Knight Riders                   98
7             Delhi Daredevils                   80
8         Kochi Tuskers Kerala                    8
9                Pune Warriors                   20

In [21]:
#data visulization of most no of toss winning team
#providing the colors to the marker line
colors =['deeppink']*15
colors[3] = 'darkmagenta'
colors[1] = 'darkmagenta'
#providing the necessary dataset and the values of x and y
fig = px.bar(data_frame=toss_winner_df , x=toss_winner_df['Team'] ,
             y=toss_winner_df["Number of Toss wins"])
# providing title 
fig.update_layout(title= 'Most number of toss winner')
#specificationns of marker line
fig.update_traces(marker_line_color='black',marker_color=colors
                  ,marker_line_width = 1.6)

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>Mumbai indians </b> has won most no of tosses(106)
</p>
</div>


<a id="33">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color: orange; color :black;">6) what does the team decide after winning the toss:</h1>


In [22]:
#what does the team decides after winning the toss
toss_decision_1  = []   #an empty list
# preparing the data of team with their decesions after the toss winning 
for decesion in matches['toss_decision'].unique():
    # appending the data into the empty list
    toss_decision_1.append({"Decision":decesion , 'No of times':len(matches[matches['toss_decision'] == decesion])})
#preparing the dataset
toss_decision_df = pd.DataFrame(data=toss_decision_1)
toss_decision_df

Decision  No of times
0    field          496
1      bat          320

In [23]:
#data visulization of decision after toss winning
#providing the required dataset and the values of x and y
fig = px.histogram(data_frame=toss_decision_df , x=toss_decision_df['Decision'],
                y=toss_decision_df['No of times'])
# providing the title
fig.update_layout(title= 'Decision After the Toss')
#specification of dataset
fig.update_traces(marker_line_color= 'black' , marker_line_width=1.7 )
fig.show()

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>Most no times the toss winning team decides fielding first</b>
</p>
</div>


<a id="33">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color: orange; color :black;">7) toss decesions across the season :</h1>


In [24]:
matches.head(2)

id        city        date player_of_match  \
0  335982   Bangalore  2008-04-18     BB McCullum   
1  335983  Chandigarh  2008-04-19      MEK Hussey   

                                        venue  neutral_venue  \
0                       M Chinnaswamy Stadium              0   
1  Punjab Cricket Association Stadium, Mohali              0   

                         team1                  team2  \
0  Royal Challengers Bangalore  Kolkata Knight Riders   
1              Kings XI Punjab    Chennai Super Kings   

                   toss_winner toss_decision                 winner result  \
0  Royal Challengers Bangalore         field  Kolkata Knight Riders   runs   
1          Chennai Super Kings           bat    Chennai Super Kings   runs   

   result_margin eliminator method    umpire1      umpire2  season  
0          140.0          N    NaN  Asad Rauf  RE Koertzen    2008  
1           33.0          N    NaN  MR Benson   SL Shastri    2008

In [25]:
# data visulizatoin of toss decisions across the season
#providing the dataset and the values of x and y
fig = px.histogram(data_frame=matches , x="season" , color='toss_decision' , barmode ='group')
# providing the title 
fig.update_layout(title= 'Toss Decisions Across The the season')
#specifications of marker line
fig.update_traces(marker_line_color='black',
                  marker_line_width=1.2, opacity=1)
fig.show()

In [26]:
# data visulization of toss decisions across the seasons
fig = px.density_heatmap(data_frame=matches , x="season" , y='toss_decision')
fig.update_layout(title= 'Toss Decisions Across The the season')
fig.show()

<a id="33">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color: orange; color :black;">8) Does wining the toss imply the wining the game :</h1>


In [27]:
#getting the data of toss winning imply the winning the game
matches['Toss_win_match_win'] = np.where(matches['toss_winner'] == matches['winner'] ,
                                             'Yes' , 'No') # using yes and no lables

In [28]:
# data visulization of toss winning imply the winning the name
#providing the data to the pie chart
fig = px.pie(values=matches['Toss_win_match_win'].value_counts(), names=['Yes','No'])
#providing the title
fig.update_layout(title='wining the toss imply the wining the game')
#marker line specifications
fig.update_traces(marker_line_width=1.2 , marker_line_color='black')
fig.show()

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>51.2 % matches syas that winning toss does imply winnning the game
</p>
</div>


<a id="33">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color: orange; color :black;">9) how many times chassing team won the toss:</h1>


In [29]:
# getting the data of tyep of win (eg. wickets or runs etc)
matches['result'].value_counts()

result
wickets    435
runs       364
tie         13
Name: count, dtype: int64

In [30]:
# getting the data of tyep of win (eg. wickets or runs etc) and converting it to the dataset
chassing_team_won = []    # empty list
#getting data of type of win
for i in  matches['result'].unique():
    #appending the data into the empty list
    chassing_team_won.append({
        'Type of win' : i , "no_of_matches" : len(matches[matches['result'] == i])
        })
# converting the data into the dta   
chassing_team_won_df =pd.DataFrame(data=chassing_team_won )
#dropping the null values
chassing_team_won_df = chassing_team_won_df.dropna()
chassing_team_won_df

Type of win  no_of_matches
0        runs            364
1     wickets            435
2         tie             13

In [31]:
#data visulization of no matches won by chassing team
# providing the dataset and the values of x and y
fig = px.bar(data_frame=chassing_team_won_df  ,y='no_of_matches', x='Type of win')
# providing the title
fig.update_layout(title= 'No. Of matches won by team while chassing and batting')
# marker specifications
fig.update_traces(marker_line_color='black',
                  marker_line_width=1.5, opacity=1)

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>435</b> Times have won the chassing team
</p>
</div>


<a id="33">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color: orange; color :black;">10) which team has won the tournament most no of times:</h1>


In [32]:
# finding out the team and their tournament titles
#data of season and winner
winning_teams = matches[['season','winner']]

# getting the data of team and their no of torunament titles
winners_team = {} #empty dictonary
for i in sorted(winning_teams['season'].unique()):
    #adding the data into the dictonary
    winners_team[i] = winning_teams[winning_teams.season == i]['winner'].tail(1).values[0] 
# sorting the data 
winners_of_IPL = pd.Series(winners_team)
#preparing the dataset
winners_of_IPL = pd.DataFrame(winners_of_IPL, columns=['team']) 

In [33]:
# team names with their no of titles
winners_of_IPL.value_counts()

team                 
Mumbai Indians           5
Chennai Super Kings      3
Kolkata Knight Riders    2
Deccan Chargers          1
Rajasthan Royals         1
Sunrisers Hyderabad      1
Name: count, dtype: int64

In [34]:
# data visu;ization to teams and their tournament title
colors =['deeppink']*15
colors[0] = 'darkmagenta'
#providing the dataset and and the values of  and y
fig = px.bar(x=winners_of_IPL.team.value_counts() , y= winners_of_IPL.team.value_counts().index 
             ,labels=dict(y='team', x='No of titles'))
# marker line specification
fig.update_traces(marker_line_color='black',
                  marker_line_width=1.5, opacity=1 , marker_color=colors)
# providing the title
fig.update_layout(title= 'No of titles won by the teams')
fig.show()

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>Mumbai Indians </b> have most IPL titles
</p>
</div>


<a id="33">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color: orange; color :black;">11) which team have played most no of games :</h1>


In [35]:
matches.head(2)

id        city        date player_of_match  \
0  335982   Bangalore  2008-04-18     BB McCullum   
1  335983  Chandigarh  2008-04-19      MEK Hussey   

                                        venue  neutral_venue  \
0                       M Chinnaswamy Stadium              0   
1  Punjab Cricket Association Stadium, Mohali              0   

                         team1                  team2  \
0  Royal Challengers Bangalore  Kolkata Knight Riders   
1              Kings XI Punjab    Chennai Super Kings   

                   toss_winner toss_decision                 winner result  \
0  Royal Challengers Bangalore         field  Kolkata Knight Riders   runs   
1          Chennai Super Kings           bat    Chennai Super Kings   runs   

   result_margin eliminator method    umpire1      umpire2  season  \
0          140.0          N    NaN  Asad Rauf  RE Koertzen    2008   
1           33.0          N    NaN  MR Benson   SL Shastri    2008   

  Toss_win_match_win  
0                 No  
1                Yes

In [36]:
# finding the no of matches played by each team
matches_played_by_teams = pd.concat([matches['team1'] ,
                                     matches['team2']]).value_counts().reset_index().rename(columns={'index' : 'Team name'})
# converting the data into the data set
matches_played_byteams_df = pd.DataFrame(data=matches_played_by_teams )
matches_played_byteams_df

Team name  count
0                Mumbai Indians    203
1   Royal Challengers Bangalore    195
2         Kolkata Knight Riders    192
3               Kings XI Punjab    190
4           Chennai Super Kings    178
5              Delhi Daredevils    161
6              Rajasthan Royals    161
7           Sunrisers Hyderabad    124
8               Deccan Chargers     75
9                 Pune Warriors     46
10               Delhi Capitals     33
11                Gujarat Lions     30
12       Rising Pune Supergiant     16
13         Kochi Tuskers Kerala     14
14      Rising Pune Supergiants     14

In [37]:
# data visulization of matches played by each team
colors = ['deeppink',] * 15
colors[0] = 'darkmagenta'
# providing the dataset and the values of x and y
fig = px.bar(data_frame=matches_played_byteams_df , x='Team name' , y='count')
#marker line specifications
fig.update_traces(marker_line_color= 'black',
                  marker_line_width= 1.5 , marker_color=colors , opacity=1)
# providing title
fig.update_layout(title= ' Number of matches played by the Team')
fig.show()


<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>Mumbai Indians</b> have played most no of matches (203)
</p>
</div>


<a id="33">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color: orange; color :black;">12) which team has won most number of matches :</h1>


In [38]:
# fist approch
# 1 using value count function 
most_wins = pd.Series(data=matches['winner'].value_counts())
#most_wins_df = pd.DataFrame(data=most_wins)
most_wins

winner
Mumbai Indians                 120
Chennai Super Kings            106
Kolkata Knight Riders           99
Royal Challengers Bangalore     91
Kings XI Punjab                 88
Rajasthan Royals                81
Delhi Daredevils                67
Sunrisers Hyderabad             66
Deccan Chargers                 29
Delhi Capitals                  19
Gujarat Lions                   13
Pune Warriors                   12
Rising Pune Supergiant          10
Kochi Tuskers Kerala             6
Rising Pune Supergiants          5
Name: count, dtype: int64

In [39]:
# second approch
# 2 using value count function  and aslo creating a data fame
most_wins = pd.concat([matches['winner']]).value_counts().reset_index().rename(columns={'index' : 'Team name'})
most_wins_byteams_df = pd.DataFrame(data=most_wins )

# creating a visulisation
colors = ['darkmagenta',] * 15
colors[0] = 'deeppink'
#providing the dataset and the values of x and y
fig = px.histogram(data_frame= most_wins_byteams_df , x='winner' , y='count')
#marker line specifications
fig.update_traces(marker_line_width= 1.5 , marker_line_color='Black' , marker_color=colors)
# title of visulisation
fig.update_layout(title= ' No of matches won by each team')

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>Mumbai Indians</b> have won most no of matches (120)
</p>
</div>


<a id="33">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color: orange; color :black;">13) which team has the highest winnig percentage:</h1>


In [40]:
# finding the no of matches played by each team 
matches_played_byteams=pd.concat([matches['team1'],matches['team2']],axis=1)
#creating data frame from team1 and team2 column
teams1=(matches_played_byteams['team1'].value_counts()+matches_played_byteams['team2'].value_counts()).reset_index()
#coulumn names
teams1.columns=['Team Name','Total Matches played']


In [41]:
#finding out the no. of wins by each team
most_wins_byteams_df=pd.DataFrame(matches['winner'].value_counts()).reset_index()
#column names
most_wins_byteams_df.columns=['Team Name','Wins']

In [42]:
#winning percentage of each team
#based on their no of wins and no of matches played by each team
#merging two dataframes to get the required data
played_matches=teams1.merge(most_wins_byteams_df,left_on='Team Name',right_on='Team Name',how='inner')
#getting winning percentage
played_matches['Win_percentage']=((played_matches['Wins']/played_matches['Total Matches played'])*100).round(2)
played_matches

Team Name  Total Matches played  Wins  Win_percentage
0           Chennai Super Kings                   178   106           59.55
1               Deccan Chargers                    75    29           38.67
2                Delhi Capitals                    33    19           57.58
3              Delhi Daredevils                   161    67           41.61
4                 Gujarat Lions                    30    13           43.33
5               Kings XI Punjab                   190    88           46.32
6          Kochi Tuskers Kerala                    14     6           42.86
7         Kolkata Knight Riders                   192    99           51.56
8                Mumbai Indians                   203   120           59.11
9                 Pune Warriors                    46    12           26.09
10             Rajasthan Royals                   161    81           50.31
11       Rising Pune Supergiant                    16    10           62.50
12      Rising Pune Supergiants                    14     5           35.71
13  Royal Challengers Bangalore                   195    91           46.67
14          Sunrisers Hyderabad                   124    66           53.23

In [43]:
#creating a visulization for highest winning percentage
#marker colors
colors = ['skyblue',] * 15
colors[11] = 'deeppink'
#providing dataframe and the values of x and y
fig = px.bar(data_frame=played_matches , x='Team Name' , y='Win_percentage')
#title
fig.update_layout(title="highest winning percentage")
#marker line specifications
fig.update_traces(marker_line_width=1.5 , marker_line_color='black' , marker_color=colors)
fig.show()

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>Rising Pune Supergiants</b> have hifhest winning percentage(62.50)
</p>
</div>


<a id="33">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color: orange; color :black;">14) lucky vanue for particular team :</h1>


In [44]:
#creating a funtion to find out the lucky vanue for particular team
def lucky(match_data, team_name, top_n=10):
    #getting the reult of each team on each vanue
    result_series = match_data[match_data['winner'] == team_name]['venue'].value_counts().nlargest(top_n)
    #converting to the dataframe and and sorting
    result_df = pd.DataFrame({'Venue': result_series.index, 'Wins': result_series.values})
    return result_df


In [45]:
# no of wins of each team on each vanue
kkr_data = lucky(matches , 'Kolkata Knight Riders')
mi_data = lucky(matches , 'Mumbai Indians')
csk_data = lucky(matches , 'Chennai Super Kings')
rcb_data = lucky(matches , 'Royal Challengers Bangalore')
kxip_data = lucky(matches , 'Kings XI Punjab')
rr_data = lucky(matches , 'Rajasthan Royals')
dd_data = lucky(matches , 'Delhi Daredevils')
Srh_data = lucky(matches , ' Sunrisers Hyderabad')
dch_data = lucky(matches , 'Deccan Chargers')
dc_data = lucky(matches , 'Delhi Capitals')
gl_data = lucky(matches , 'Gujarat Lions')
pw_data = lucky(matches , 'Pune Warriors')
rps_data = lucky(matches , 'Rising Pune Supergiant')
ktk_data = lucky(matches , ' Kochi Tuskers Kerala')
    

In [46]:
#creating a data visulization for Kolkata Knight Riders
fig = px.pie(kkr_data, values=kkr_data['Wins'], names=kkr_data['Venue'],
             title='No of wins at each vanune Kolkata Knight Riders played')
fig.update_traces(textposition='inside', textinfo='percent+label')
fig.show()

#creating a data visulization for mumbai indians
fig = px.pie(mi_data, values=mi_data['Wins'], names=mi_data['Venue'],
             title='No of wins at each vanune Mumbai indians played')
fig.update_traces(textposition='inside', textinfo='percent+label' )
fig.show()

#creating a data visulization for chennai supar kings
fig = px.pie(csk_data, values=csk_data['Wins'], names=csk_data['Venue'],
             title='No of wins at each vanune Chennai Super Kings played')
fig.update_traces(textposition='inside', textinfo='percent+label')
fig.show()

#creating a data visulization for royal challengers benglore
fig = px.pie(rcb_data, values=rcb_data['Wins'], names=rcb_data['Venue'],
             title='No of wins at each vanune Royal Challengers Bangalore Played')
fig.update_traces(textposition='inside', textinfo='percent+label')
fig.show()


# same kind of procedure can be done to data visulization of other teams

<a id="33">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color: orange; color :black;">15) inning wise comparision between teams :</h1>


<a id="33">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 20px; font-style: normal; background-color: skyblue; color :black;">15.1) Batting wise comaparision :</h1>


In [47]:
# finding out the runs of each team in scored in first and second inning
runs_df=deleviries.groupby(['id','inning','batting_team'])[['total_runs']].sum().reset_index()
#dropping unnecessary columns
runs_df.drop('id',axis=1,inplace=True)

#batting wise compirision (batting first and second batting)
Batting_first = runs_df[runs_df['inning'] == 1] #1st inning runs
Batting_second = runs_df[runs_df['inning'] == 2] #2nd inning runs

In [48]:
#creating data visulization for batting first
#providing the dataframe and the values of x,y and color
fig = px.box(data_frame=Batting_first, y='total_runs',x='batting_team',color='batting_team')
# title and font size
fig.update_layout(title='Inning wise comparision Batting_first',titlefont={'size': 25})
fig.show()

##creating data visulization for bowling second
#providing the dataframe and the values of x,y and color
fig = px.box(data_frame=Batting_second, y='total_runs',x='batting_team',color='batting_team')
#title and font size
fig.update_layout(title='Inning wise comparision Batting_second',titlefont={'size': 25})
fig.show()


<a id="33">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 20px; font-style: normal; background-color: skyblue; color :black;">15.2) Bowling wise comaparision:</h1>


In [49]:
#gtting the data of wickets in each match
wicket_df=deleviries.groupby(['id','inning','bowling_team'])[['is_wicket']].sum().reset_index()
#dropping unnecessary data(columns)
wicket_df.drop('id',axis=1,inplace=True)

#bowling wise compirision (bowling first and bowling second)
Bowling_first = wicket_df[wicket_df['inning'] == 1] # 1st inning
bowling_second = wicket_df[wicket_df['inning'] == 2] # 2nd inning

In [50]:
#creating data visulization for bowling first
#providing the dataframe and the values of x,y and color
fig = px.box(data_frame=Bowling_first, y='is_wicket',x='bowling_team',color='bowling_team')
#title and font size
fig.update_layout(title='Inning wise comparision Bowling_first',titlefont={'size': 25})
fig.show()

##creating data visulization for bowling second
#providing the dataframe and the values of x,y and color
fig = px.box(data_frame=bowling_second, y='is_wicket',x='bowling_team',color='bowling_team')
#title and font size
fig.update_layout(title='Inning wise comparision bowling_second',titlefont={'size': 25})
fig.show()


<a id="33">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color: orange; color :black;">16) which team has scored most no. of 200+ :</h1>


In [51]:
#finding out which team hsa scored most no of 200+ runs
#groupping the data from deliveries dataset and calculacting the total runs of each team in each match
score =deleviries.groupby(['id','inning','batting_team' , 'bowling_team'])[['total_runs']].sum().reset_index()
#checking total runs 
score_df = score[score['total_runs'] >= 200]
#counting no of 200+ runs 
team_1=score_df['batting_team'].value_counts()
#dataframe
team_1=pd.DataFrame(team_1)

In [52]:
#creating a visuliztion for the no of 200+ runs scored by an team
#providiing colors to the marker
colors = ['turquoise',] * 11
colors[0] = 'darkmagenta'
# providing the required data and the values of x and y
fig=px.bar(x=team_1.index,y=team_1['count'],labels=dict(x="Team Name",y="Count"),)
# titale and font size
fig.update_layout(title="Total count of 200+ by batting team",
                  titlefont={'size': 26}     
                  )
#marker line specifications
fig.update_traces(marker_line_color='black',
                  marker_line_width=2, opacity=1,marker_color=colors)
fig.show()

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>Royal Challengers Bangalore</b> scored most 200+ scores.
</p>
</div>


<a id="33">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color: orange; color :black;">17) which team has conceded most no. of 200+ :</h1>


In [53]:
#finding out the which team has conceded most 200+ scores
#using the data which we have obtained for most 200+
#only diierence is here we are calculating the most 200+ conceded
conceded_200 = score_df['bowling_team'].value_counts()
# dataframe
conceded_200 = pd.DataFrame(conceded_200)

#provinfing colors to the marker
colors = ['turquoise',] * 11
colors[0] = 'darkmagenta'
#providing the required data and values of x and y 
fig = px.bar(data_frame=conceded_200 , x=conceded_200.index , 
             y=score_df['bowling_team'].value_counts() , labels=dict(x='Bowling Team' , y='count of 200+ conceded'),)
#title and font size
fig.update_layout(title="Total count of 200+ conceded by batting team",
                  titlefont={'size': 26}     
                  )
#marker line specification
fig.update_traces(marker_line_color='black',
                  marker_line_width=2, opacity=1,marker_color=colors)
fig.show()

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>Kings XI Punjab</b> conceded most 200+ scores.
</p>
</div>


<a id="33">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color: orange; color :black;">18) Highest score by a team in a inning :</h1>


In [54]:
# finding which team has the highest score in ea single inning
#groupping the data and calculating total runs of each team in each match
score =deleviries.groupby(['id','inning','batting_team' , 'bowling_team'])[['total_runs']].sum().reset_index()
#score_df = score[score['total_runs'] >= 200]
#printing the highest score in a single inning
print('HIghest score by a team ' ,':', score['total_runs'] .max())

HIghest score by a team  : 263


<div class="alert alert-block alert-info"><p style='color:black;'> 📌
HIghest score by a team  : 263 runs.<br> 
</p>
</div>

<a id="30">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color:orange; color :black;">19) biggest win margin in terms of runs :</h1>



In [55]:
# printing the biggest win in terms of runs
print( 'Biggest win margin in terms of runs', ':',matches['result_margin'].max())

Biggest win margin in terms of runs : 146.0


<div class="alert alert-block alert-info"><p style='color:black;'> 📌
Biggest win margin in terms of runs : 146 runs.<br> 
</p>
</div>

<a id="30">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color:orange; color :black;">20) Which batsman have played most no of balls :</h1>



In [56]:
# finding which batsman has played most no of balls
#groupping the data and finding the batsman name and no balls played by him
most_balls_played_bybatsmen = deleviries.groupby(['batsman'])['ball'].count().reset_index()
#sorting the data based on no balls played by batsman
most_ball_played_bybatsmen=most_balls_played_bybatsmen.sort_values(by='ball',ascending=False).head(30)
#most_ball_played_bybatsmen

In [57]:
# creating visulization for no of balls played by the batsman
x1 = most_ball_played_bybatsmen['batsman'] #batsman name
y1 = most_ball_played_bybatsmen['ball']    #balls played by the batsman

#provifiing the dataframe and the values of x and y
fig = px.line(data_frame=most_ball_played_bybatsmen , x=x1, y=y1 )
#title and font size
fig.update_layout(title="No of balls played by the batsmen",
                  titlefont={'size': 25}     
                  )
fig.show()

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>Virat kohli</b> have palyed : 4609 balls.
</p>
</div>


<a id="30">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color:orange; color :black;">21) Leading run scorer of all time </h1>



In [58]:
#finding leading run scorer of all time
#groupping the data and geeting the total runs of batsman
leading_run_scorrer = deleviries.groupby(['batsman'])['batsman_runs'].sum().reset_index()
#sorting based on no of runs
leading_run_scorrer=leading_run_scorrer.sort_values(by='batsman_runs',ascending=False).head(30)

In [59]:
#creating a visulization for leading run scorer of all time
batsmen = leading_run_scorrer['batsman'] #batsman name
total_runs = leading_run_scorrer['batsman_runs'] #batsman runs

#providing the data amd values of x and y
fig = px.bar(data_frame=most_ball_played_bybatsmen , x=batsmen, y=total_runs ,color=most_ball_played_bybatsmen['ball']
             ,labels=dict( x='batsmen', y='total_runs' ))
#title and font size
fig.update_layout(title="Leading run scorrer of all time with no. of balls played",
                  titlefont={'size': 20}     
                  )
#marker line specifications
fig.update_traces(marker_line_color='black',
                  marker_line_width=1.5, opacity=1)
fig.show()

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>Virat kohli</b> is leading run scorer of all time with 5878 runs
</p>
</div>


<a id="30">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color:orange; color :black;">22) Who has hit most no of 4s </h1>


In [60]:
#most no of fours by the batsman
#getting the 4s data
fours=deleviries[deleviries['batsman_runs']==4]
#groupping the data and counting the no of fours
most_4s_bybatsmen = fours.groupby(['batsman'])['batsman_runs'].value_counts().reset_index()
#sorting
most_4s_bybatsmen_df = most_4s_bybatsmen.sort_values(by='count' , ascending=False).head(20)

In [61]:
#creating a visulization for most no of fours by the batsman
colors = ['mediumpurple',] * 30
colors[0] = 'midnightblue'
#providing the dataframe and the values of x and with labels
fig = px.bar(data_frame=most_4s_bybatsmen_df , x='batsman',y='count'
             ,labels=dict(x='batsmen', y='total_4s'),)
#title
fig.update_layout(title="Most no of 4s hit by a Batsmen")
#marker line specifications
fig.update_traces(marker_line_color='black',
                  marker_line_width=1.5, opacity=1,marker_color=colors)
fig.show()

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>S Dhawan</b> have the record of hitting most fours (591).
</p>
</div>


<a id="30">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color:orange; color :black;">23) Who has hit most no of 6s </h1>


In [62]:
#most no of 6s by the batsman
#getting the 6s data
Sixs=deleviries[deleviries['batsman_runs']== 6]
#groupping the data and getting the no of 6s
most_6s_bybatsmen = Sixs.groupby(['batsman'])['batsman_runs'].value_counts().reset_index()
#sorting
most_6s_bybatsmen_df = most_6s_bybatsmen.sort_values(by='count' , ascending=False).head(20)

In [63]:
#creating a visulization for most no of 6s
colors = ['deepskyblue',] * 30
colors[0] = 'midnightblue'
colors[18] = 'grey'
colors[19] = 'grey'
#providing the data and values of x and y with labels
fig = px.bar(data_frame=most_6s_bybatsmen_df , x='batsman',y='count'
             ,labels=dict(x='batsmen', y='total_6s'),)
#title
fig.update_layout(title="Most no of 6s hit by a Batsmen")
#marker line specifications
fig.update_traces(marker_line_color='black',
                  marker_line_width=1.5, opacity=1,marker_color=colors)
fig.show()

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>CH Gayle</b> have the record of hitting most Six's (349).
</p>
</div>


<a id="30">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color:orange; color :black;">24) Who has the highest strike rate </h1>


In [64]:
#finding the highest strike rate 
#groupping the data and no of balls played by the batsman
strike_rate = deleviries.groupby(['batsman'])['ball'].count().reset_index()
#sorting
highest_strike_rate = strike_rate.sort_values(by='ball',ascending=False).head(20)
#geetting strike by dividing batsman runs and no balls played by him
st_rate = ((leading_run_scorrer['batsman_runs'] / highest_strike_rate['ball'])*100).round(2)
#new column
highest_strike_rate['st_rate'] = st_rate


In [65]:
#strike rate dataframe
highest_strike_rate.head()

batsman  ball  st_rate
505    V Kohli  4609   127.53
407   S Dhawan  4208   123.50
379  RG Sharma  4088   127.94
438   SK Raina  4041   132.84
116  DA Warner  3819   137.58

In [66]:
#soting
highest_strike_rate=pd.DataFrame(data=highest_strike_rate.sort_values(by='st_rate' , ascending=False))

In [67]:
#creating a visulization for highest strike rate 
colors = ['turquoise',] * 30
colors[0] = 'midnightblue'
colors[1] = 'midnightblue'
#providing a dataframe and the the values of x and y
fig = px.bar(data_frame=highest_strike_rate , x='batsman' , y='st_rate')
#title
fig.update_layout(title=" Who has the highest strike rate ")
# marker line specifications
fig.update_traces(marker_line_color='black',
                  marker_line_width=1.5, opacity=1,marker_color=colors)
fig.show()

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>AB de Villiers</b> have the highest strike rate with a strike rate of 148.56.
</p>
</div>


<a id="30">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color:orange; color :black;">25) Leading wicket taker </h1>


In [68]:
#finding leadding wicket taker by groupping the data
wicket_taker = deleviries.groupby(['bowler']) ['is_wicket'].sum().reset_index()
#sorting
leading_wicket_taker = wicket_taker.sort_values(by='is_wicket' , ascending=False).head(20)

In [69]:
# creating a visulization for leading wicket taker 
colors = ['lightyellow',] * 30
colors[0] = 'turquoise'
colors[1] = 'turquoise'
colors[18] = 'grey'
colors[19] = 'grey'
#providing the dataset and values of x and y
fig = px.bar(data_frame=leading_wicket_taker , x='bowler' , y='is_wicket')
#title
fig.update_layout(title=" Leading wicket taker ")
#marker line specifications
fig.update_traces(marker_line_color='black',
                  marker_line_width=1.5, opacity=1,marker_color=colors)
fig.show()

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>SL Malinga</b> have the record of taking most wickets (188).
</p>
</div>


<a id="30">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color:orange; color :black;">26) Which stadium hosted most no. of games </h1>


In [70]:
#stadium hosted most no of games
#groupping data and finding no of matches at each vanue
matches_each_vanue = matches.groupby(['venue'])['venue'].value_counts().reset_index()
#sorting
no_of_matches_each_vanue = matches_each_vanue.sort_values(by='count' , ascending=False)#.head(10)
#new column
no_of_matches_each_vanue['No of matches']= no_of_matches_each_vanue['count']
#dropping column
no_of_matches_each_vanue.drop('count' , axis=1 , inplace=True)
no_of_matches_each_vanue.head()

venue  No of matches
7                                Eden Gardens             77
8                            Feroz Shah Kotla             74
35                           Wankhede Stadium             73
14                      M Chinnaswamy Stadium             65
24  Rajiv Gandhi International Stadium, Uppal             64

In [71]:
#visulization for most no games at particular vanue
#providing the dataframe and values of x and y
fig = px.pie(data_frame=no_of_matches_each_vanue , names='venue' , values='No of matches')
fig.update_traces(textposition='inside', textinfo='percent+label')
#title
fig.update_layout(title = 'Stadium hosted no of games')
fig.show()

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
Above pie chart indicates that till 2020 Iden Garden hosted 9.44 % of . matches which is 77 matches  .<br> 
</p>
</div>

<a id="30">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color:orange; color :black;">27) Who has won most no. of MOTM </h1>


In [72]:
#calculating whos has won most no motm by gropping the data
MOtM = matches.groupby(['player_of_match'])['player_of_match'].value_counts().reset_index()
#sorting
no_of_MOTM = MOtM.sort_values(by='count' , ascending=False).head(15)
#new column
no_of_MOTM['No_of_MOTM']= no_of_MOTM['count']
#dropping column
no_of_MOTM.drop('count' , axis=1 , inplace=True)
no_of_MOTM.head(5)

player_of_match  No_of_MOTM
10   AB de Villiers          23
35         CH Gayle          22
173       RG Sharma          18
44        DA Warner          17
138        MS Dhoni          17

In [73]:
#visulization for player with no of MOTM
colors = ['lavenderblush',] * 15
colors[0] = 'navy'
colors[1] = 'navy'

#providing datdframe and values of x and y
fig = px.histogram(data_frame=no_of_MOTM , x='player_of_match' , y= 'No_of_MOTM' )
#title
fig.update_layout(title=" Most no of Man of the match ")
#marker line specifications
fig.update_traces(marker_line_color='black',
                  marker_line_width=1.5, opacity=1,marker_color=colors)
fig.show()

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>AB de Villiers</b> hav won most no of MOTM (23 Times).
</p>
</div>


<a id="30">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color:orange; color :black;">28) Count of fours hit in each season </h1>


In [74]:
#counting no of fours in each season
season_val = matches['season'].unique()
fours_count = [] #wmpty list
for season in season_val:
    #checking
    season_df = matches[matches['season']==season]
    total_fours = 0
    for i in season_df['id'].values:
        temp_df = deleviries[deleviries['id']==i]
        #checking
        fours = temp_df[temp_df['batsman_runs']==4]['batsman_runs'].count()
        total_fours+=fours
        #appending the data
    fours_count.append(total_fours)

In [75]:
#visulization for no of fours in each season
colors = ['moccasin',] * 15
colors[5] = 'deeppink'
colors[3] = 'deeppink'
#providing the dataframe and values of x and y
fig = px.bar(x=season_val , y=fours_count , labels=dict(x='Season' , y='No. of fours'))
#title
fig.update_layout(title='No. of fours in each season')
#specifications
fig.update_traces(marker_color=colors, marker_line_width=1.2 ,
                  marker_line_color="black")

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>2013</b> season has most 4s of all time (2052 fours).
</p>
</div>


<a id="30">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color:orange; color :black;">29) Count of six hit in each season </h1>


In [76]:
#no of 6s in each season
season_val = matches['season'].unique()
Sixs_count = [] #empty list
for season in season_val:
    #checking
    season_df = matches[matches['season']==season]
    total_six = 0
    for i in season_df['id'].values:
        #checking 
        temp_df = deleviries[deleviries['id']==i]
        sixs = temp_df[temp_df['batsman_runs']==6]['batsman_runs'].count()
        total_six+=sixs
        #appending the data
    Sixs_count.append(total_six)

In [77]:
#visulization for no of 6s in each season
colors = ['deeppink',] * 15
colors[10] = 'moccasin'
colors[11] = 'moccasin'
#values of x and y with dataframe
fig = px.bar(x=season_val , y=Sixs_count , labels=dict(x='Season' , y='No. of Sixs'))
#title
fig.update_layout(title='No. of Sixs in each season')
#marker line specifications
fig.update_traces(marker_color=colors, marker_line_width=1.5 ,
                  marker_line_color="black")

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>2018</b> season has most 6s of all time (872 six's).
</p>
</div>


<a id="30">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color:orange; color :black;">30) Runs scored from boundries in each season </h1>


In [78]:
runs_4=np.dot(fours_count,4)   #dot product of runs of 4
runs_6=np.dot(Sixs_count,6)       #dot product of runs of 6

In [79]:
#creating a visulization for runs scored with boundries in each season
#providing the values of x and y
fig = px.bar(x=matches['season'].unique() , y=(runs_4+runs_6 ) , labels=dict(x='season', y='Runs_scored'))
#marker line colors
colors = ['deeppink',] * 15
colors[4] = 'moccasin'
colors[5] = 'moccasin'
#title
fig.update_layout(title='Runs scored from boundries in each season')
#marker line specifications
fig.update_traces(marker_color=colors, marker_line_width=1.5 ,
                  marker_line_color="black")
fig.show()

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>12.258k</b> runs were scored in 2013 season
</p>
</div>


<a id="30">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color:orange; color :black;">31) Runs scored from fours in each season </h1>


In [80]:
#finding runs scored from fours in a each season
runs_byfours =[]         #empty list
for i in fours_count: #taking values from fours
    runs_byfours.append(i*4)  #multiplying each value by 4
#marker colors  
colors = ['moccasin',] * 15
colors[5] = 'deeppink'
colors[3] = 'deeppink'
 #providing the dataframe and values of x and y with labels
fig= px.bar(x=season_val , y=runs_byfours , labels=dict(x='season' , y='Runs'))
#title
fig.update_layout(title=' Runs scored from fours in each season')
#marker line specifications
fig.update_traces(marker_color=colors, marker_line_width=1.2 ,
                  marker_line_color="black")

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>8208</b> runs scored in 2013 season.
</p>
</div>


<a id="30">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color:orange; color :black;">32) which team has scored the most runs in first 6 overs  </h1>


In [81]:
#team has scored most runs in first 6 overs
#unique names of team
team_name = deleviries['batting_team'].unique()
team_runs = []  #empty list
for i in team_name:
    #checking deliviries team unique name is same or not if same then getting dataframe
    tmp_df = deleviries[deleviries['batting_team']==i]
    #checking the overs 
    tmp_df = tmp_df[tmp_df['over'].isin([0,1,2,3,4,5])]
    runs = tmp_df['total_runs'].sum()
    #appending the data 
    team_runs.append(runs)

#converting to the dataframe
team_runs_in_first6over = pd.DataFrame(data=team_runs, index=team_name,columns=['Runs In First 6 Overs'])
team_runs_in_first6over.index.name = 'Team'


In [82]:
team_runs_in_first6over.head()

Runs In First 6 Overs
Team                                              
Kolkata Knight Riders                         8871
Royal Challengers Bangalore                   8699
Kings XI Punjab                               8907
Chennai Super Kings                           8048
Rajasthan Royals                              7151

In [83]:
#creating a visulization for runs in first 6 overs
#maeker colors
colors = ['turquoise',] * 15
colors[6] = 'moccasin'
 #providing the dataframe and values of x and y with labels
fig= px.bar(y=team_runs_in_first6over['Runs In First 6 Overs'] ,
            x=team_runs_in_first6over.index , labels=dict(x='Team name' , y='Runs'))
#title
fig.update_layout(title='Runs scored in first 6 overs')
#marker line specifications
fig.update_traces(marker_color=colors, marker_line_width=1.5 ,
                  marker_line_color="black")

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>Mumbai Indians</b> scored 9204 runs in first 6 overs.
</p>
</div>


<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color:orange; color :black;">33) which team has scored the most runs in last 4 overs  </h1>


In [84]:
#team has scored most runs in last 4 overs
#unique names of team
team_name_1 = deleviries['batting_team'].unique()
team_runs_4 = []
for i in team_name_1:
    #checking deliviries team unique name is same or not if same then getting dataframe
    tmp_df = deleviries[deleviries['batting_team']==i]
    #checking the overs
    tmp_df = tmp_df[tmp_df['over'].isin([16,17,18,19])]
    runs = tmp_df['total_runs'].sum()
    #appending the data
    team_runs_4.append(runs)
    
#converting to the dataframe
team_name_1 = pd.DataFrame(data=team_runs_4, index=team_name_1,columns=['Runs In last 4 Overs'])
team_name_1.index.name = 'Team'
#sorting
team_name_1 = team_name_1.sort_values(by='Runs In last 4 Overs', ascending=False)

In [85]:
team_name_1.head()

Runs In last 4 Overs
Team                                             
Mumbai Indians                               7538
Royal Challengers Bangalore                  6787
Chennai Super Kings                          6655
Kings XI Punjab                              6227
Kolkata Knight Riders                        6224

In [86]:
colors = ['turquoise',] * 15
colors[0] = 'moccasin'
 #providing the dataframe and values of x and y with labels
fig= px.bar(y=team_name_1['Runs In last 4 Overs'] ,
            x=team_name_1.index , labels=dict(x='Team name' , y='Runs'))
#title
fig.update_layout(title='Runs scored in last 4 overs')
#marker line specifications
fig.update_traces(marker_color=colors, marker_line_width=1.5 ,
                  marker_line_color="black")

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>Mumbai Indians</b> scored 7538 runs in last 4 overs.
</p>
</div>


<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color:orange; color :black;">34) which team has the best scoring run_rate in first 6 overs  </h1>


In [87]:
#checking scoring run rate in first 6 overs
#merging data to get new dataframe with runs in first 6 overs
played_df=played_matches.merge(team_runs_in_first6over,left_on='Team Name',right_on='Team',how='inner')
#merging data to get new dataframe with runs in last 4 overs
played_df2=played_df.merge(team_name_1,left_on='Team Name',right_on='Team',how='inner')
#calculating matches
Runrate=np.dot(played_df2['Total Matches played'],6)
Runrate1=np.dot(played_df2['Total Matches played'],4)
#calculating run rate
played_df2['Run-Rate in first 6 overs']=((played_df2['Runs In First 6 Overs']/Runrate)).round(2)
played_df2['Run-Rate in last 4 overs']=((played_df2['Runs In last 4 Overs']/Runrate1)).round(2)
played_df2.head()

Team Name  Total Matches played  Wins  Win_percentage  \
0  Chennai Super Kings                   178   106           59.55   
1      Deccan Chargers                    75    29           38.67   
2       Delhi Capitals                    33    19           57.58   
3     Delhi Daredevils                   161    67           41.61   
4        Gujarat Lions                    30    13           43.33   

   Runs In First 6 Overs  Runs In last 4 Overs  Run-Rate in first 6 overs  \
0                   8048                  6655                       7.54   
1                   3417                  2539                       7.59   
2                   1560                  1160                       7.88   
3                   7349                  5043                       7.61   
4                   1553                   921                       8.63   

   Run-Rate in last 4 overs  
0                      9.35  
1                      8.46  
2                      8.79  
3                      7.83  
4                      7.68

In [88]:
#creating a visulization for  run rate in first 6 overs
#marker colors
colors = ['moccasin',] * 15
colors[4] = 'deeppink'
colors[11] = 'deeppink'
#providing dataframe and values of x and y
fig = px.histogram(data_frame=played_df2 , x='Team Name' , y='Run-Rate in first 6 overs')
#title
fig.update_layout(title="Run-Rate in Last 6 overs")
#marker line specifications
fig.update_traces(marker_color=colors , marker_line_color='Black' , marker_line_width=1.2)
fig.show()

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>Gujarat Lions</b> has highest run rate in first 6 overs(8.63).
</p>
</div>


<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; font-style: normal; background-color:orange; color :black;">35) which team has the best scoring run_rate  in last 4 overs  </h1>


In [89]:
#creating a visulization for scoring run rate in last 4 overs
#marker colors
colors = ['moccasin',] * 15
colors[0] = 'deeppink'
colors[8] = 'deeppink'
#providing dataframe and values of x and y
fig = px.histogram(data_frame=played_df2 , x='Team Name' , y='Run-Rate in last 4 overs')
#title
fig.update_layout(title="Run-Rate in Last 6 overs")
#marker line specifications
fig.update_traces(marker_color=colors , marker_line_color='Black' , marker_line_width=1.2)
fig.show()

<div class="alert alert-block alert-info"><p style='color:black;'> 📌
<b>Chennai Supar Kings</b> has highest run rate in last 4 overs(9.35).
</p>
</div>


<a id="33">
<h1 style = "font-family: Snell Roundhand;font-weight: 500;font-size: 30px; text-align: center; font-style: normal; background-color: orange; color :black;"> Thank You..... </h1>
